packages/react-dom/src/server/ReactPartialRenderer.js

```diff
@@ -776,6 +777,16 @@ class ReactDOMServerRenderer {
     context[this.threadID] = previousValue;
   }
 
+  clearProviders(): void {
+    while (this.contextIndex > -1) {
+      if (__DEV__) {
```

Comment:
Can you write the loop inline with minimal work it needs to do? For example reading previous value seems unnecessary except for the last one. This would also avoid the overloading you had to add to `popProvider`.

Reply:
I've pushed another commit which does as you request.

However, the previous value does need to be restored for each Provider, as each Context's current value is stored separately on that Context object.

Well, if there were several nested Providers for the same Context, then yes you could only restore just the last previous value. But I don't think this would be more performative, as `clearProviders()` would need to loop over the array twice - once to identify any repeated contexts, and then a 2nd time to restore whatever was the oldest previous value for each Context.

I think it's probably cheaper to redundantly write the value multiple times in these cases.

But if you want further changes, please just say so and I'll implement.

Reply:
NB This leaves `contextStack` and `contextValueStack` unaltered, and so retaining memory. But I guess since the `PartialRenderer` instance has been destroyed, it will be dereferenced and garbage collected shortly anyway.